File: LeetCode_C++/0062.不同路径.cpp
/*
 * @lc app=leetcode.cn id=62 lang=cpp
 *
 * [62] 不同路径
 */

// @lc code=start
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for(int i = 0; i < m; i++)
        {
            for(int j = 0; j < n; j++)
            {
                if(!i && !j) dp[i][j] = 1;  // 左上角
                if(i) dp[i][j] += dp[i - 1][j];  // 从上面过来
                if(j) dp[i][j] += dp[i][j - 1];  // 从左边过来
            }
        }
        return dp[m - 1][n - 1];
    }
};
// @lc code=end

File: LeetCode_Java/0062.不同路径.java
/*
 * @lc app=leetcode.cn id=62 lang=java
 *
 * [62] 不同路径
 */

// @lc code=start
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] f = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (i == 1 && j == 1) f[i][j] = 1;
                else f[i][j] = f[i - 1][j] + f[i][j - 1];
            }
        }
        return f[m][n];
    }
}
// @lc code=end

